Update README for 1.0.0

kriskowal opened the pull request.

abhinav commented on README.md:
````diff
@@ -1,26 +1,227 @@
+and YARPC bindings for the echo service.
+
+```go
+//go:generate thriftrw --plugin=yarpc echo.thrift
````
We should get them to install the plugin as well if we're going to tell them to use this command

abhinav commented on README.md:
````diff
@@ -1,26 +1,227 @@
+
+```go
+ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+defer cancel()
````
This reminds me: We should document somewhere that contexts must have a timeout or we will reject the request.

kriskowal replied:
noted

abhinav commented on README.md:
```diff
@@ -1,26 +1,227 @@
 
-## Development Status: Alpha
+Unstable:
+- Any package in an `x` directory, including the experimental Redis transport
```
Inconsistent capitalization

kriskowal replied:
Redis is a proper noun, whereas round-robin is a mere noun.

abhinav commented on README.md:
```diff
@@ -1,26 +1,227 @@
+  Chooser, List
+- the middleware API
+- wire representation of RPC for HTTP and TChannel, including all required
+  headers: Rpc-Caller, Rpc-Service, Rpc-Procedure, and Context-TTL.
```
Context-TTL-MS

akshayjshah commented on README.md:
```diff
@@ -1,26 +1,227 @@
+choosers.
+YARPC for Go provides reference implementations for HTTP/1.1 and TChannel
+transports, and also raw, JSON, and Thrift encodings.
+YAPRC for Go provides experimental implementations for a Redis transport and a
```
spelling: `YARPC`.

Maybe we should add lint checks for common misspellings (like `YAPRC`)?

akshayjshah commented on README.md:
````diff
@@ -1,26 +1,227 @@
 ```
+signals := make(chan os.Signal, 1)
+signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+<-signals
````
Should this perhaps be a method on the `Dispatcher` type or a small standalone function? I'd love it if users didn't need to worry about which signals they should trap.

```go
func (d *Dispatcher) Notify() <-chan os.Signal {
  ch := make(chan os.Signal, 1)
  signal.Notify(ch, syscall.SIGINT, syscall.SIGTERM)
  return ch
}

func main() {
  ...set up everything...
  defer dispatcher.Stop()
  <-dispatcher.Notify()
}
```

kriskowal replied:
I believe the dispatcher is a convenient but inappropriate home for this solution, particularly since it is the wrong place to iterate on this solution when an application grows out of it.

akshayjshah replied:
That's probably true. Would providing a small standalone function be better? It could even be something as opaque as `yarpc.WaitForShutdownSignal()`, which completely hides the channel.

akshayjshah commented on README.md:
````diff
@@ -1,26 +1,227 @@
+```go
+type helloHandler struct{}
+
+func (h helloHandler) Echo(ctx context.Context, e *echo.EchoRequest) (*echo.EchoResponse, error) {
````
Pointer receiver to match line 131 above?

akshayjshah commented on README.md:
```diff
@@ -1,26 +1,227 @@
+
+To send a request on an outbound, we construct a client using the corresponding
+named outbound from the dispatcher.
+The client will use that name for the outbound request service name
```
nit: terminal punctuation.

blampe commented:
A point of confusion I'm seeing is that people think the encodings in here are prescriptive (you can only use these) instead of extensible (define your own!). I think it would help to have a small example of how to extend a handler to use gob, for example, if that's small enough to fit in here.

kriskowal commented:
@blampe This change includes a TODO list for example topics. I’ve added a note to add an example of how to create a custom encoding. Gob fits the bill well, or we can create a pared down JSON encoding for example purposes only. I also added a note that we can add examples for custom transports (maybe simple socket using an envelope), peer chooser (maybe based on sharding), peer lists (stripped down round robin), and peer providers (DNS), to really bring the point home that YARPC is interfaces and a couple reference implementations.

breerly commented on README.md:
```diff
@@ -1,26 +1,242 @@
-A transport and encoding agnostic RPC framework.
+With hundreds to thousands of services communicating with RPC, transport
+protocols (like HTTP and [TChannel][]), encoding protocols (like JSON or
+Thrift), and peer choosers are the concepts that vary.
```
suggest "...choosers are concepts that vary constantly from year to year."

breerly commented on README.md:
```diff
@@ -1,26 +1,242 @@
+With hundreds to thousands of services communicating with RPC, transport
+protocols (like HTTP and [TChannel][]), encoding protocols (like JSON or
+Thrift), and peer choosers are the concepts that vary.
+Separating these concerns allows services to change transport protocols without
```
The terminology here is odd - suggest either just "transports" or "wire protocols" - these are mostly interchangeable, with transports implying a accompanying library that *implements* the wire protocol. Perhaps we should put this in a terminology doc.

breerly commented on README.md:
```diff
@@ -1,26 +1,242 @@
 
-If using [Glide](https://github.com/Masterminds/glide), *at least* `glide version 0.12` is required to install:
+If using [Glide](https://github.com/Masterminds/glide), *at least* `glide
+version 0.12` is required to install:
```
This should be `0.12.3` actually - anything prior doesn't work with semver.

breerly commented on README.md:
```diff
@@ -1,26 +1,242 @@
 glide version 0.12.2
 
-$ glide get 'go.uber.org/yarpc#~0.5'
+$ glide get 'go.uber.org/yarpc#^1'
```
^1 - that is correct :)

breerly commented on README.md:
````diff
@@ -1,26 +1,242 @@
+```
+
+
+## Examples
````
I would rather link directly to the example dirs, where we can add `README` to each one.

kriskowal replied:
I agree about linking. I’ve added a link to this particular example. I think it’s still useful to embed a guide inline for at least one to give skimmers something to evaluate the project without linking through.
